active_discussion/embedded/static-initialization.md

```diff
@@ -0,0 +1,137 @@
+  accepted and implemented.
+
+[rfc2585]: https://github.com/rust-lang/rfcs/pull/2585
+
+## Memory clobber synchronization
+
+Consider this program
+
+``` rust
+#![no_std]
+
+use core::mem::MaybeUninit;
+
+static mut X: MaybeUninit<bool> = MaybeUninit::uninit();
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    X.write(false);
+
+    unsafe {
+        asm!("ENABLE_INTERRUPTS" : : : "memory" : "volatile");
+        //                             ^^^^^^^^
+    }
+
+    // `INTERRUPT0` can preempt `main` from this point on and at any time
+
+    loop {
+        // .. any safe code ..
+    }
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut bool = unsafe { &mut *X.as_mut_ptr() };
+
+    // .. any safe code ..
+}
+```
+
+Note that "any safe code" can *not* call `main` or `INTERRUPT0` (because they
+are `unsafe` functions), use `asm!` or access registers.
+
+**Claim**: the memory clobber is sufficient to prevent misoptimizations.
+
+"Why is the memory clobber required?" If the compiler reorders `X.write` to
+after the `asm!` block `INTERRUPT0` could observe `X` in an uninitialized state.
+
+## Compiler fence synchronization
+
+Consider this program which is a slight variation of the first one:
+
+``` rust
+#![no_std]
+
+use core::{
+    cell::UnsafeCell,
+    mem::MaybeUninit,
+    ptr,
+    sync::atomic::{self, Ordering},
+};
+
+extern "C" {
+    static MASK_INTERRUPT: UnsafeCell<u8>;
+    static UNMASK_INTERRUPT: UnsafeCell<u8>;
+}
+
+const ORDERING: Ordering = ..;
+
+static mut X: MaybeUninit<bool> = MaybeUninit::uninit();
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        // mask INTERRUPT0
+        ptr::write_volatile(MASK_INTERRUPT.get(), 1 << 0);
+        asm!("ENABLE_INTERRUPTS" : : : : "volatile");
+    }
+
+    X.write(false);
+
+    atomic::compiler_fence(ORDERING);
+
+    unsafe {
+        // unmask INTERRUPT0
+        ptr::write_volatile(UNMASK_INTERRUPT.get(), 1 << 0);
+    }
+
+    // `INTERRUPT0` can preempt `main` from this point on and at any time
+
+    loop {
+        // .. any safe code ..
+    }
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut bool = unsafe { &mut *X.as_mut_ptr() };
+
+    // .. any safe code ..
+}
+```
+
+**Claim**: the compiler fence is sufficient to prevent misoptimization provided
+that `ORDERING` is any of: `Ordering::SeqCst` `Ordering::AcqRel` or
+`Ordering::Release`.
+
+"Why is the compiler fence required?" if the compiler reorders `X.write` to
+after the second `write_volatile` statement `INTERRUPT0` could observe `X` in an
+uninitialized state.
+
+## Questions
+
+- Can these programs be misoptimized by the compiler? For example, if the
+  compiler optimizes away the `X.write(..)` statement `INTERRUPT0` would observe
+  an `X` in an uninitialized state. Is it allowed to do that?
+
+  - If yes, what additional constraints are required to prevent misoptimization?
+    Should `X` be initialized using a volatile write?
```

Comment:
It doesn't need to be.  Just for the record:
- If you did use a volatile write and a volatile read, it would remove the need for the `memory` clobber or `compiler_fence`, because the purpose of those in this case is to flush pending regular writes.  (The `asm!` would still require `volatile` because it has side effects). 
- Using a volatile write but not a volatile read would be incorrect per Rust docs: [Memory accessed with `read_volatile` or `write_volatile` should not be accessed with non-volatile operations.](https://doc.rust-lang.org/nightly/std/ptr/fn.write_volatile.html).  I believe it would be legal in terms of LLVM semantics, though.

Reply:
> If you did use a volatile write and a volatile read, it would remove the need for the memory clobber or compiler_fence,

I am not sure about this. Volatile accesses are ordered with respect to each other, but here you assume they are also ordered wrt. whatever will make it so that "interrupts can only happen after this point".

> Using a volatile write but not a volatile read would be incorrect per Rust docs: "Memory accessed with read_volatile or write_volatile should not be accessed with non-volatile operations." I believe it would be legal in terms of LLVM semantics, though.

That's a *should*. It basically just means "that's likely incorrect and you should re-think what you are doing".